### ISSUE TITLE
Failed to install app cluster-monitoring

### ISSUE THREAD

**Issue by andrezaycev** [rancher/rancher#21078]
**What kind of request is this (question/bug/enhancement/feature request):**

Failed cluster-monitoring bug

**Steps to reproduce (least amount of steps as possible):**

Enable monitoring cluster level

**Result:**
If check Systems/apps - cluster-monitoring have installing state

**Other details that may be helpful:**

```
Failed to install app cluster-monitoring. Error: release cluster-monitoring failed: Prometheus.monitoring.coreos.com "cluster-monitoring" is invalid: []: Invalid value: map[string]interface {}{"apiVersion":"monitoring.coreos.com/v1", "kind":"Prometheus", "metadata":map[string]interface {}{"name":"cluster-monitoring", "namespace":"cattle-prometheus", "creationTimestamp":"2019-06-24T08:49:14Z", "labels":map[string]interface {}{"chart":"prometheus-0.0.1", "heritage":"Tiller", "io.cattle.field/appId":"cluster-monitoring", "release":"cluster-monitoring", "app":"prometheus"}, "generation":1, "uid":"f17bab4a-965c-11e9-a3c7-00259052e531"}, "spec":map[string]interface {}{"additionalScrapeConfigs":map[string]interface {}{"name":"prometheus-cluster-monitoring-additional-scrape-configs", "key":"additional-scrape-configs.yaml"}, "affinity":map[string]interface {}{"podAntiAffinity":map[string]interface {}{"preferredDuringSchedulingIgnoredDuringExecution":[]interface {}{map[string]interface {}{"podAffinityTerm":map[string]interface {}{"labelSelector":map[string]interface {}{"matchLabels":map[string]interface {}{"app":"prometheus", "prometheus":"cluster-monitoring"}}, "topologyKey":"kubernetes.io/hostname"}, "weight":100}}}}, "baseImage":"rancher/prom-prometheus", "containers":[]interface {}{map[string]interface {}{"name":"prometheus-proxy", "ports":[]interface {}{map[string]interface {}{"containerPort":80, "name":"http", "protocol":"TCP"}}, "resources":map[string]interface {}{"limits":map[string]interface {}{"cpu":"100m", "memory":"100Mi"}, "requests":map[string]interface {}{"cpu":"50m", "memory":"50Mi"}}, "securityContext":map[string]interface {}{"runAsNonRoot":false, "runAsUser":0}, "volumeMounts":[]interface {}{map[string]interface {}{"name":"configmap-prometheus-cluster-monitoring-nginx", "mountPath":"/nginx"}}, "command":[]interface {}{"/bin/sh", "-c", "cp /nginx/run-sh.tmpl /var/run/nginx-start.sh; chmod +x /var/run/nginx-start.sh; /var/run/nginx-start.sh"}, "env":[]interface {}{map[string]interface {}{"valueFrom":map[string]interface {}{"fieldRef":map[string]interface {}{"fieldPath":"status.podIP"}}, "name":"POD_IP"}}, "image":"rancher/nginx:1.15.8-alpine"}, map[string]interface {}{"env":[]interface {}{map[string]interface {}{"name":"POD_IP", "valueFrom":map[string]interface {}{"fieldRef":map[string]interface {}{"fieldPath":"status.podIP"}}}}, "image":"rancher/prometheus-auth:v0.2.0", "livenessProbe":map[string]interface {}{"httpGet":map[string]interface {}{"path":"/-/healthy", "port":"web", "scheme":"HTTP"}, "periodSeconds":10, "successThreshold":1, "timeoutSeconds":10, "failureThreshold":6}, "ports":[]interface {}{map[string]interface {}{"containerPort":9090, "name":"web", "protocol":"TCP"}}, "args":[]interface {}{"--proxy-url=http://127.0.0.1:9090", "--listen-address=$(POD_IP):9090", "--filter-reader-labels=prometheus", "--filter-reader-labels=prometheus_replica"}, "command":[]interface {}{"prometheus-auth"}, "name":"prometheus-agent", "readinessProbe":map[string]interface {}{"periodSeconds":10, "successThreshold":1, "timeoutSeconds":10, "failureThreshold":10, "httpGet":map[string]interface {}{"path":"/-/ready", "port":"web", "scheme":"HTTP"}}, "resources":map[string]interface {}{"limits":map[string]interface {}{"cpu":"500m", "memory":"200Mi"}, "requests":map[string]interface {}{"memory":"100Mi", "cpu":"100m"}}}}, "evaluationInterval":"60s", "externalLabels":map[string]interface {}{"prometheus_from":"adp1"}, "logFormat":interface {}(nil), "replicas":1, "retention":"12h", "scrapeInterval":"60s", "configMaps":[]interface {}{"prometheus-cluster-monitoring-nginx"}, "logLevel":"info", "nodeSelector":map[string]interface {}{"beta.kubernetes.io/os":"linux"}, "ruleSelector":map[string]interface {}{"matchExpressions":[]interface {}{map[string]interface {}{"key":"source", "operator":"In", "values":[]interface {}{"rancher-alert", "rancher-monitoring"}}}}, "serviceAccountName":"cluster-monitoring", "serviceMonitorNamespaceSelector":map[string]interface {}{"matchExpressions":[]interface {}{map[string]interface {}{"operator":"In", "values":[]interface {}{"p-fzvhd"}, "key":"field.cattle.io/projectId"}, map[string]interface {}{"operator":"In", "values":[]interface {}{"p-fzvhd"}, "key":"field.cattle.io/projectId"}}}, "serviceMonitorSelector":map[string]interface {}{"matchExpressions":[]interface {}{}}, "additionalAlertManagerConfigs":map[string]interface {}{"key":"additional-alertmanager-configs.yaml", "name":"prometheus-cluster-monitoring-additional-alertmanager-configs"}, "listenLocal":true, "securityContext":map[string]interface {}{"runAsNonRoot":true, "runAsUser":1000, "fsGroup":2000}, "version":"v2.7.1", "paused":false, "podMetadata":map[string]interface {}{"labels":map[string]interface {}{"app":"prometheus", "chart":"prometheus-0.0.1", "release":"cluster-monitoring"}}, "resources":map[string]interface {}{"limits":map[string]interface {}{"cpu":"1000m", "memory":"1000Mi"}, "requests":map[string]interface {}{"cpu":"750m", "memory":"750Mi"}}, "ruleNamespaceSelector":map[string]interface {}{"matchExpressions":[]interface {}{map[string]interface {}{"values":[]interface {}{"p-fzvhd"}, "key":"field.cattle.io/projectId", "operator":"In"}, map[string]interface {}{"key":"field.cattle.io/projectId", "operator":"In", "values":[]interface {}{"p-fzvhd"}}}}, "secrets":[]interface {}{"exporter-etcd-cert"}}}: 

validation failure list: spec.logFormat in body must be of type string: "null"

```

**Environment information**
- Rancher version (`rancher/rancher`/`rancher/server` image tag or shown bottom left in the UI): 2.2.4
- Installation option (single install/HA): single

**Cluster information**
- Cluster type (Hosted/Infrastructure Provider/Custom/Imported): Custom
- Machine type (cloud/VM/metal) and specifications (CPU/memory): 3 machine, 8 cpu 16 GBRAM
- Kubernetes version (use `kubectl version`): 

```
1.13.5
```

- Docker version (use `docker version`):

```
18.09.2
```

**Comment by Serrvosky** [comment]
I'm having the same issue here

**Comment by loganhz** [comment]
@guangbochen

**Comment by guangbochen** [comment]
You need to disable the monitoring and re-enable with set the `prometheus.logFormat=logfmt` in the `show advanced options` section to address this issue temporarily.

![image](https://user-images.githubusercontent.com/4569037/60063675-4d24ea00-9730-11e9-8a38-a595afa2f793.png)
This should be a legacy issue, as the `prometheus.logFormat` is not required in the [previous CRD spec](https://github.com/coreos/prometheus-operator/blob/master/Documentation/api.md#prometheusspec), but it is required for now, have you update the CRD resource manually or installed community version in the same cluster?

**Comment by loganhz** [comment]
Please also see https://github.com/rancher/rancher/issues/20769#issuecomment-501095945 if you have prometheus operator deployed before 

We will fix it in 2.3. For now, please try to use the workload in https://github.com/rancher/rancher/issues/20769#issuecomment-501095945 or https://github.com/rancher/rancher/issues/21078#issuecomment-505248052

**Comment by andrezaycev** [comment]
> Please also see [#20769 (comment)](https://github.com/rancher/rancher/issues/20769#issuecomment-501095945) if you have prometheus operator deployed before
> 
> We will fix it in 2.3. For now, please try to use the workload in [#20769 (comment)](https://github.com/rancher/rancher/issues/20769#issuecomment-501095945) or [#21078 (comment)](https://github.com/rancher/rancher/issues/21078#issuecomment-505248052)

Thank you. Delete crd and redploy cluster agent help for me.

**Comment by jiaqiluo** [comment]
Tried the following steps but did not get the bug to reproduce successfully:

- run Rancher:v2.2.6
- add a cluster
- go to `System` project -> `Apps`, launch the app `Prometheus` from the built-in catalog library
- run the kubectl to see that the following CRDs created
```
alertmanagers.monitoring.coreos.com           2019-07-31T21:51:56Z
prometheuses.monitoring.coreos.com            2019-07-31T21:51:56Z
prometheusrules.monitoring.coreos.com         2019-07-31T21:51:56Z
servicemonitors.monitoring.coreos.com         2019-07-31T21:51:56Z
```
- delete the app `Prometheus`
- enable the cluster monitoring 

Result:
- cluster monitoring is enabled successfully

**Comment by jiaqiluo** [comment]
The bug is reproduced in Rancher v2.2.6

Steps:
- run Rancher:v2.2.6
- add a cluster
- enable the helm stable catalog
- launch the app `promethuese-operator` from helm stable catalog into the `kube-system` namespace in `System` project
- delete the app after it is active
- enable the cluster monitoring 

Result:
- the app `cluster-monitoring` show the following error:
```
Failed to install app cluster-monitoring. Error: release cluster-monitoring failed: Prometheus.monitoring.coreos.com "cluster-monitoring" is invalid: []: Invalid value: map[string]interface {}{"apiVersion":"monitoring.coreos.com/v1", "kind":"Prometheus", "metadata":map[string]interface {}{"creationTimestamp":"2019-08-01T03:17:30Z", "generation":1, "labels":map[string]interface {}{"app":"prometheus", "chart":"prometheus-0.0.1", "heritage":"Tiller", "io.cattle.field/appId":"cluster-monitoring", "release":"cluster-monitoring"}, "name":"cluster-monitoring", "namespace":"cattle-prometheus", "uid":"e5cf7c71-b40a-11e9-84d4-3eea20bf883f"}, "spec":map[string]interface {}{"additionalAlertManagerConfigs":map[string]interface {}{"key":"additional-alertmanager-configs.yaml", "name":"prometheus-cluster-monitoring-additional-alertmanager-configs"}, "additionalScrapeConfigs":map[string]interface {}{"key":"additional-scrape-configs.yaml", "name":"prometheus-cluster-monitoring-additional-scrape-configs"}, "affinity":map[string]interface {}{"podAntiAffinity":map[string]interface {}{"preferredDuringSchedulingIgnoredDuringExecution":[]interface {}{map[string]interface {}{"podAffinityTerm":map[string]interface {}{"labelSelector":map[string]interface {}{"matchLabels":map[string]interface {}{"app":"prometheus", "prometheus":"cluster-monitoring"}}, "topologyKey":"kubernetes.io/hostname"}, "weight":100}}}}, "baseImage":"rancher/prom-prometheus", "configMaps":[]interface {}{"prometheus-cluster-monitoring-nginx"}, "containers":[]interface {}{map[string]interface {}{"command":[]interface {}{"/bin/sh", "-c", "cp /nginx/run-sh.tmpl /var/run/nginx-start.sh; chmod +x /var/run/nginx-start.sh; /var/run/nginx-start.sh"}, "env":[]interface {}{map[string]interface {}{"name":"POD_IP", "valueFrom":map[string]interface {}{"fieldRef":map[string]interface {}{"fieldPath":"status.podIP"}}}}, "image":"rancher/nginx:1.15.8-alpine", "name":"prometheus-proxy", "ports":[]interface {}{map[string]interface {}{"containerPort":80, "name":"http", "protocol":"TCP"}}, "resources":map[string]interface {}{"limits":map[string]interface {}{"cpu":"100m", "memory":"100Mi"}, "requests":map[string]interface {}{"cpu":"50m", "memory":"50Mi"}}, "securityContext":map[string]interface {}{"runAsNonRoot":false, "runAsUser":0}, "volumeMounts":[]interface {}{map[string]interface {}{"mountPath":"/nginx", "name":"configmap-prometheus-cluster-monitoring-nginx"}}}, map[string]interface {}{"args":[]interface {}{"--proxy-url=http://127.0.0.1:9090", "--listen-address=$(POD_IP):9090", "--filter-reader-labels=prometheus", "--filter-reader-labels=prometheus_replica"}, "command":[]interface {}{"prometheus-auth"}, "env":[]interface {}{map[string]interface {}{"name":"POD_IP", "valueFrom":map[string]interface {}{"fieldRef":map[string]interface {}{"fieldPath":"status.podIP"}}}}, "image":"rancher/prometheus-auth:v0.2.0", "livenessProbe":map[string]interface {}{"failureThreshold":6, "httpGet":map[string]interface {}{"path":"/-/healthy", "port":"web", "scheme":"HTTP"}, "periodSeconds":10, "successThreshold":1, "timeoutSeconds":10}, "name":"prometheus-agent", "ports":[]interface {}{map[string]interface {}{"containerPort":9090, "name":"web", "protocol":"TCP"}}, "readinessProbe":map[string]interface {}{"failureThreshold":10, "httpGet":map[string]interface {}{"path":"/-/ready", "port":"web", "scheme":"HTTP"}, "periodSeconds":10, "successThreshold":1, "timeoutSeconds":10}, "resources":map[string]interface {}{"limits":map[string]interface {}{"cpu":"500m", "memory":"200Mi"}, "requests":map[string]interface {}{"cpu":"100m", "memory":"100Mi"}}}}, "evaluationInterval":"60s", "externalLabels":map[string]interface {}{"prometheus_from":"do"}, "listenLocal":true, "logFormat":interface {}(nil), "logLevel":"info", "nodeSelector":map[string]interface {}{"beta.kubernetes.io/os":"linux"}, "paused":false, "podMetadata":map[string]interface {}{"labels":map[string]interface {}{"app":"prometheus", "chart":"prometheus-0.0.1", "release":"cluster-monitoring"}}, "replicas":1, "resources":map[string]interface {}{"limits":map[string]interface {}{"cpu":"1000m", "memory":"1000Mi"}, "requests":map[string]interface {}{"cpu":"750m", "memory":"750Mi"}}, "retention":"12h", "ruleNamespaceSelector":map[string]interface {}{"matchExpressions":[]interface {}{map[string]interface {}{"key":"field.cattle.io/projectId", "operator":"In", "values":[]interface {}{"p-j69t9"}}, map[string]interface {}{"key":"field.cattle.io/projectId", "operator":"In", "values":[]interface {}{"p-j69t9"}}}}, "ruleSelector":map[string]interface {}{"matchExpressions":[]interface {}{map[string]interface {}{"key":"source", "operator":"In", "values":[]interface {}{"rancher-alert", "rancher-monitoring"}}}}, "scrapeInterval":"60s", "secrets":[]interface {}{"exporter-etcd-cert"}, "securityContext":map[string]interface {}{"fsGroup":2000, "runAsNonRoot":true, "runAsUser":1000}, "serviceAccountName":"cluster-monitoring", "serviceMonitorNamespaceSelector":map[string]interface {}{"matchExpressions":[]interface {}{map[string]interface {}{"key":"field.cattle.io/projectId", "operator":"In", "values":[]interface {}{"p-j69t9"}}, map[string]interface {}{"key":"field.cattle.io/projectId", "operator":"In", "values":[]interface {}{"p-j69t9"}}}}, "serviceMonitorSelector":map[string]interface {}{"matchExpressions":[]interface {}{}}, "version":"v2.7.1"}}: validation failure list: 
spec.logFormat in body must be of type string: "null"
```

**Comment by jiaqiluo** [comment]
The bug fix is validated in Rancher:master-head ef42cc4

following the same steps as the previous comment, the cluster monitoring works.
